### PR TITLE
Assign groups to be graded to TA's, fixes #397 (work in progress)

### LINKED ISSUE #397: Assign groups to be graded to TA's

### PR DESCRIPTION
**Pull request by JWGmeligMeyling**
Work in progress, opened PR for code review.

### CODE REVIEW

File: src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    public AssignedTAs(EntityManager em) {
+        super(em);
+    }
+
+    @Transactional
+    public List<Delivery> getLastDeliveries(Assignment assignment, User user) {
+        Map<Group, List<Delivery>> deliveriesMap = query().from(delivery)
+                .where(delivery.assignment.eq(assignment)
+                        .and(delivery.group.groupNumber.in(
+                                new JPASubQuery()
```
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java]: Perhaps we can break this subquery out of the query builder and put it into a variable or function that creates the subquery (`groupNumbersForAssistantByAssignmentSubQuery(..., ...)` because this get's pretty hard to read.

File: src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import static com.mysema.query.group.GroupBy.groupBy;
+import static com.mysema.query.group.GroupBy.list;
+import static java.util.Comparator.comparing;
+import static nl.tudelft.ewi.devhub.server.database.entities.QAssignedTA.assignedTA;
+import static nl.tudelft.ewi.devhub.server.database.entities.QDelivery.delivery;
+
+/**
+ * Created by sayra on 07/06/2017.
+ */
+public class AssignedTAs extends Controller<AssignedTA> {
```
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java]: Even though the `AssignedTAs` class is probably hit in the end to end test suite, it still may be an idea to write an integration test that checks whether the relatively complex query defined in `AssignedTAs#getLastDerliveries(Assignment, User)` actually works (or at least executes). See also how the other controllers/ Data Access Objects are tested at https://github.com/devhub-tud/devhub/tree/master/src/test/java/nl/tudelft/ewi/devhub/server/database/controllers .

File: src/main/java/nl/tudelft/ewi/devhub/server/database/entities/AssignedTA.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import javax.persistence.*;
+import javax.validation.constraints.NotNull;
+import java.util.List;
+
+/**
+ * Created by Pilmus on 31-5-2017.
+ */
+@Data
+@Entity
+@Table(name = "assigned_ta")
```
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/entities/AssignedTA.java]: Could use a [`@UniqueConstraint`](http://docs.oracle.com/javaee/6/api/javax/persistence/UniqueConstraint.html) on `course_edition_id`, `group_number` and `assignment`.

File: src/main/java/nl/tudelft/ewi/devhub/server/database/entities/Assignment.java [hunk 4]
```diff
@@ -97,50 +98,61 @@
...
-	}
-
-	public List<Task> copyTasksFromOldAssignment(Assignment oldAssignment) {
-		return oldAssignment.getTasks().stream().map(this::taskforNewAssginment).collect(Collectors.toList());
-	}
-
-	private Task taskforNewAssginment(Task oldTask) {
-		return oldTask.copyForNextYear(this);
-	}
+    @JsonIgnore
```
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/entities/Assignment.java]: Try to avoid reformatting big unchanged chunks of code. They can cause a lot of difficulty when merging pull requests or figuring out what change introduced a particular problem. Usually its best to revert the change if you see auto formatting was applied prior to committing.

File: src/main/java/nl/tudelft/ewi/devhub/server/web/resources/AssignmentsResource.java [hunk 5]
```diff
@@ -298,8 +302,21 @@ public Response getAssignmentPage(@PathParam("courseCode") String courseCode,
...
         Assignment assignment = assignmentsDAO.find(course, assignmentId);
+
+        List<Delivery> userDeliveries = assignedTAs.getLastDeliveries(assignment, currentUser);
         List<Delivery> lastDeliveries = deliveriesDAO.getLastDeliveries(assignment);
-        AssignmentStats assignmentStats = deliveriesBackend.getAssignmentStats(assignment, lastDeliveries);
+
+
+
+        AssignmentStats assignmentStats;
+        if(currentUser.isAdmin()) {
```
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/web/resources/AssignmentsResource.java]: We have discussed the option of making this toggle'able. Is that still on?

File: src/main/resources/changelog.xml [hunk 6]
```diff
@@ -1,296 +1,297 @@
...
 
 <databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
-	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
-	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.0.xsd
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.0.xsd
         http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd">
 
-	<changeSet id="create_users_table" author="Michael de Jong">
```
Comment (JWGmeligMeyling) [inline src/main/resources/changelog.xml]: I am afraid that reformatting in the changelog.xml may cause a lot of trouble, so I'll reapply the changes when I'll be merging this.

File: src/main/resources/changelog.xml [hunk 7]
```diff
@@ -934,182 +986,236 @@
...
+
+        <addForeignKeyConstraint constraintName="user_ref_in_assigned_ta"
+                                 baseTableName="assigned_ta" baseColumnNames="ta_id"
+                                 referencedTableName="users" referencedColumnNames="id"/>
+
+        <addForeignKeyConstraint constraintName="course_ref_in_assigned_ta"
+                                 baseTableName="assigned_ta" baseColumnNames="assignment_course_edition_id"
+                                 referencedTableName="assignments" referencedColumnNames="course_edition_id"/>
+
+        <addForeignKeyConstraint constraintName="assignment_ref_in_assigned_ta"
```
Comment (JWGmeligMeyling) [inline src/main/resources/changelog.xml]: Nice constraints! 👍 Could perhaps use a `unique constraint` on `course_edition_id`, `group_number` and `assignment_id` too (to prevent multiple TA's being assigned, or the same TA being assigned twice).

File: src/main/java/nl/tudelft/ewi/devhub/server/database/entities/Assignment.java [hunk 8]
```diff
@@ -43,48 +43,49 @@
...
     @Column(name = "assignment_id")
-	@GeneratedValue(generator = "seq_group_number", strategy = GenerationType.AUTO)
-	@GenericGenerator(name = "seq_group_number", strategy = "nl.tudelft.ewi.devhub.server.database.entities.identity.FKSegmentedIdentifierGenerator", parameters = {
-		@org.hibernate.annotations.Parameter(name = FKSegmentedIdentifierGenerator.TABLE_PARAM, value = "seq_assignment_id"),
-		@org.hibernate.annotations.Parameter(name = FKSegmentedIdentifierGenerator.CLUSER_COLUMN_PARAM, value = "course_edition_id"),
-		@org.hibernate.annotations.Parameter(name = FKSegmentedIdentifierGenerator.PROPERTY_PARAM, value = "courseEdition")
-	})
+    @GeneratedValue(generator = "seq_group_number", strategy = GenerationType.AUTO)
+    @GenericGenerator(name = "seq_group_number", strategy = "nl.tudelft.ewi.devhub.server.database.entities.identity.FKSegmentedIdentifierGenerator", parameters = {
+            @org.hibernate.annotations.Parameter(name = FKSegmentedIdentifierGenerator.TABLE_PARAM, value = "seq_assignment_id"),
```
Comment (Fastjur) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/entities/Assignment.java]: Did you auto reformat this? There are a lot of changed lines here that are simply whitespace. Was this intentional?

File: src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java (hunk 2 above)
Comment (sranjha) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java]: Testing the query is indeed a good idea. I'll have a look at this after isolating the subquery since testing the queries separately would be more convenient.

File: src/main/resources/changelog.xml (hunk 7 above)
Comment (sranjha) [inline src/main/resources/changelog.xml]: Done!

File: src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java (hunk 1 above)
Comment (pilmus) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java]: Agreed!

File: src/main/java/nl/tudelft/ewi/devhub/server/database/entities/Assignment.java (hunk 4 above)
Comment (pilmus) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/entities/Assignment.java]: My apologies, I'm in the habit of using IntelliJs reformatting shortcuts quite frequently. Will pay more attention in the future!

File: src/main/java/nl/tudelft/ewi/devhub/server/web/resources/AssignmentsResource.java (hunk 5 above)
Comment (pilmus) [inline src/main/java/nl/tudelft/ewi/devhub/server/web/resources/AssignmentsResource.java]: Yup :)

File: src/main/java/nl/tudelft/ewi/devhub/server/database/entities/Assignment.java (hunk 8 above)
Comment (pilmus) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/entities/Assignment.java]: Yes I reformatted it, and no it wasn't intentional; JW pointed it out in another comment as well. Will avoid in the future!

File: src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
+package nl.tudelft.ewi.devhub.server.database;
```
Comment (pilmus) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java]: Could we do something where each TA has a field indicating how many groups they can be assigned to? I can imagine that not each TA has an equal amount of time for grading for example.

File: src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Lists.partition;
+import static java.util.Collections.shuffle;
+
+/**
+ * Created by sayra on 07/06/2017.
+ */
+public class AssignTAs {
+    public static List<AssignedTA> assignGroups(Collection<User> TAs, Collection<Group> groupList, Assignment assignment, Random random) {
```
Comment (pilmus) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java]: Do I understand it correctly that this method will randomly distribute the groups over the TAs every time it is called? If so, we might want to add some code that prevents the assignments to be shuffled each time the method is invoked.

File: src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java (hunk 9 above)
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java]: We could definitely add that as an enhancement but I think foremost it makes sense to be able to reassign an assigned group to someone else.

File: src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java (hunk 10 above)
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java]: Yes that would happen, but we did thought of that, where this method is used, its only invoked on the groups that have not been assigned yet for this assignment.

File: src/main/java/nl/tudelft/ewi/devhub/server/database/entities/Assignment.java (hunk 8 above)
Comment (Fastjur) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/entities/Assignment.java]: Alright no problem!

File: src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java (hunk 9 above)
Comment (TimvdLippe) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java]: Note that we actually do this and it is a requirement for this year. Manual reassignment is possible, but not desirable. CC @mjwsteenbergen as he is the original author of the assignment script we are using atm.

File: src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        List<AssignedTA> assignedTAs = new ArrayList<>();
+
+        List<Group> groups = new ArrayList<>();
+        groups.addAll(groupList);
```
Comment (TimvdLippe) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java]: You can do this directly with the [`Collection` constructor](https://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html#ArrayList(java.util.Collection))

```java
List<Group> groups = new ArrayList<>(groupList);
```

**Comment by TimvdLippe** [comment]
You can more easily review this PR with https://github.com/devhub-tud/devhub/pull/475/files?w=1 (note the `?w=1` appended to the url)

File: src/main/java/nl/tudelft/ewi/devhub/server/web/resources/AssignmentsResource.java [hunk 12]
```diff
@@ -728,4 +737,33 @@ public Assignment updateAssignment(@PathParam("courseCode") String courseCode,
...
+
+        List<Delivery> deliveries = deliveriesDAO.getLastDeliveries(assignment);
+
+        List<Group> groups = Lists.transform(deliveries, Delivery::getGroup);
+
+        Set<User> TAs = course.getAssistants();
+
+        groups.removeAll(Lists.transform(assignment.getAssignedTAS(), AssignedTA::getGroup));
+
+        List<AssignedTA> assignedTAS = AssignTAs.assignGroups(TAs, groups, assignment, new Random());
```
Comment (TimvdLippe) [inline src/main/java/nl/tudelft/ewi/devhub/server/web/resources/AssignmentsResource.java]: Why do you construct `Random` outside of `assignGroups`?

File: src/test/java/nl/tudelft/ewi/devhub/server/backend/Bootstrapper.java [hunk 13]
```diff
@@ -297,6 +297,16 @@ private void prepareGroup(Map<String, User> userMapping, CourseEdition entity, M
...
 			deliveries.persist(deliveryEntity);
 
+			AssignedTA assignedTA = new AssignedTA();
+			assignedTA.setAssignment(deliveryEntity.getAssignment());
+			assignedTA.setGroup(groupEntity);
+			if(deliveryCounter % 2 == 0) {
+				assignedTA.setTeachingAssistant(userMapping.get("assistant1"));
+				entityManager.persist(assignedTA);
+			} else {
+				//assignedTA.setTeachingAssistant(userMapping.get("assistant2"));
```
Comment (TimvdLippe) [inline src/test/java/nl/tudelft/ewi/devhub/server/backend/Bootstrapper.java]: Add a `TODO` here to indicate this still needs to be fixed (or fix it directly in this PR)

File: src/test/java/nl/tudelft/ewi/devhub/server/database/AssignTAsTest.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    List<Group> singleGroup;
+    List<User> teachingAssistants;
+    Assignment assignment;
+    AssignedTA assignedTA1Group2;
+    AssignedTA assignedTA1Group3;
+    AssignedTA assignedTA1Group1;
+    AssignedTA assignedTA1Group4;
+    AssignedTA assignedTA2Group1;
+    AssignedTA assignedTA2Group4;
+    List<AssignedTA> assignedTAS;
```
Comment (TimvdLippe) [inline src/test/java/nl/tudelft/ewi/devhub/server/database/AssignTAsTest.java]: Constructing these in a `setUp` and them only use them once or twice in a test hurts readibility imo. Would prefer to just construct them in the test directly. That way the test is concise and contains all the information you need. Personally I use fields when I use them in almost all the tests and the setup for the value is complicated/tedious (for example to set up stubs)

File: src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java (hunk 9 above)
Comment (mjwsteenbergen) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java]: I think that it would be a good idea to implement this in the base feature, as current TA's are expecting this feature to be there as I have already built it. (Sorry about that)
My implementation can be found in this [gist]( https://gist.github.com/mjwsteenbergen/4e29c62a63d72525317e07ab51349b21)

**Comment by Fastjur** [comment]
Man I wish my PRs were so full of discussion when I was doing the star module...

File: src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java (hunk 9 above)
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java]: Thanks for linking your code 👍 Actually I quickly jumped by the TA desk to ask for it but you were already gone. We'll see if we can include it but its beyond the scope of this PR and we will evaluate when we will make this enhancement!

File: src/main/java/nl/tudelft/ewi/devhub/server/web/resources/AssignmentsResource.java (hunk 12 above)
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/web/resources/AssignmentsResource.java]: Doesn't make a lot of sense to pass this as a parameter instead of through dependency injection, but I haven't explained DI for now as we have been focussing on other technologies for now. Passing it as a parameter still makes more sense than instantiating it in the method itself, it indicates that a Random is used and that if preferred, a more secure type of random can be provided.

File: src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java (hunk 9 above)
Comment (TimvdLippe) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java]: Luckily the next deadline is in 2 weeks, so you have some time :smile: Not sure if we will use the DevHub feature if this enhancement is not included, as the gist already does the work and it is crucial to our TA assignment. If you want to add it in a different PR or now, I don't mind either way :stuck_out_tongue:

**Comment by TimvdLippe** [comment]
Oh and one more enhancement, which is probably not a strict requirement, is to assign a TA to a group iff that TA did not grade that group already for a previous assignment. Not sure how complicated that enhancement would be. Our current gist does not take this into account, so we can live without it :P

File: src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java (hunk 11 above)
Comment (sranjha) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java]: That's better indeed. I changed it

File: src/test/java/nl/tudelft/ewi/devhub/server/database/AssignTAsTest.java (hunk 14 above)
Comment (sranjha) [inline src/test/java/nl/tudelft/ewi/devhub/server/database/AssignTAsTest.java]: I moved the construction of AssignedTAs to their respective tests to improve readability

File: src/main/resources/static/js/deliveries-filter.js [hunk 15]
```diff
@@ -8,6 +8,8 @@ $(function() {
 		var deliveries = $('tr.delivery');
 		var table = deliveries.closest("table");
 
+		console.log(table);
```
Comment (TimvdLippe) [inline src/main/resources/static/js/deliveries-filter.js]: Left-over debug statement

File: src/main/resources/static/js/toggle-view.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
+/**
+ * Created by Maaike Visser on 13-6-2017.
+ */
+
+$(function () {
+    "use strict";
+
+    var view = true;
```
Comment (TimvdLippe) [inline src/main/resources/static/js/toggle-view.js]: Instead of a boolean, I think it is easier to use [`toggle()`](http://api.jquery.com/toggle/)

File: src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java (hunk 2 above)
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java]: I've added a test for the query.

File: src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java (hunk 9 above)
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/AssignTAs.java]: Discussed with Liam, we will not support TA based capacities just yet because it causes imbalance in invested time across TA's. Element of this decision is that it takes a bit of time to implement a proper UI for this that also ensures that there is enough capacity to distribute all groups, and that all team members currently have more important issues to work on. We definitely do want to support this feature at some point though, we're just not sure if all requirements are clear right now and how to implement it right.

Instead we will allow TA's to handle the situation themselves and pass on assigned groups to someone else. Therefore we added the ability to reassign a group.

File: src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java (hunk 1 above)
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java]: Done!

File: src/test/java/nl/tudelft/ewi/devhub/server/backend/Bootstrapper.java (hunk 13 above)
Comment (JWGmeligMeyling) [inline src/test/java/nl/tudelft/ewi/devhub/server/backend/Bootstrapper.java]: Actually this has to be removed. And I think if groups are assigned to assistants anyway, this should be done through the JSON configuration and not in the bootstrapper itself.

File: src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java (hunk 2 above)
Comment (sranjha) [inline src/main/java/nl/tudelft/ewi/devhub/server/database/controllers/AssignedTAs.java]: Nice!
